File: day-2/puzzle.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Game map[string]int

func ParseGame(line string) []Game {
	parts := strings.SplitN(line, ":", 2)
	_, subsets := parts[0], parts[1]

	// var gameId int
	// _, err := fmt.Sscanf(gameIdStr, "Game %d", &gameId)
	// if err != nil {
	// 	panic(err)
	// }

	subsetsMaps := []Game{}
	for _, subset := range strings.Split(subsets, ";") {
		subsetMap := Game{}
		for _, cubes := range strings.Split(subset, ",") {
			var num int
			var color string
			fmt.Sscanf(cubes, "%d %s", &num, &color)
			subsetMap[color] = num
		}
		subsetsMaps = append(subsetsMaps, subsetMap)
	}
	return subsetsMaps
}

func MaxValues(games []Game) Game {
	max := Game{}
	for _, game := range games {
		for color, num := range game {
			if max[color] < num {
				max[color] = num
			}
		}
	}
	return max
}

func main() {
	file, err := os.Open(os.Args[1])
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	gameId := 0
	possibleIdsSum := 0
	powerSum := 0
	for scanner.Scan() {
		gameId++
		line := scanner.Text()
		games := ParseGame(line)
		// fmt.Printf("%v\n", games)

		maxGame := MaxValues(games)
		// fmt.Printf("Max values: %v\n", maxGame)

		if maxGame["red"] <= 12 && maxGame["green"] <= 13 && maxGame["blue"] <= 14 {
			possibleIdsSum += gameId
			fmt.Printf("Game %d possible\n", gameId)
		}
		power := maxGame["red"] * maxGame["green"] * maxGame["blue"]
		fmt.Printf("Game %d power: %d\n", gameId, power)
		powerSum += power
	}
	fmt.Printf("Possible games id sum: %d\n", possibleIdsSum)
	fmt.Printf("Sum of powers: %d\n", powerSum)
}

File: day-2/puzzle.lua
#!/usr/bin/env lua

function Main()
	if #arg < 1 then
		error("Not enough arguments provided, need one with the input filename")
	end

	local file = io.open(arg[1], "r")
	if not file then
		error("Could not open file" .. arg[1])
	end

	--- @type string
	local content = file:read("*a"):gsub("^%s+", ""):gsub("%s+$", "")
	file:close()

	local sumOfIds = 0
	local gameId = 0
	local powerSum = 0
	for line in content:gmatch("[^\r\n]+") do
		gameId = gameId + 1
		local _, endPos = line:find(": ", 0, true)
		local cubesStr = line:sub(endPos + 1)

		---@type table<string,number>
		local gameMax = {
			red = 0,
			green = 0,
			blue = 0,
		}

		for game in cubesStr:gmatch("[^;]+") do
			for ball in (game .. ", "):gmatch("(.-), ") do
				local numberS, color = ball:match("(%d+) (%w+)")
				local num = tonumber(numberS)
				print(num, color)
				if gameMax[color] < num then
					gameMax[color] = num
				end
			end
		end
		if gameMax.red <= 12 and gameMax.green <= 13 and gameMax.blue <= 14 then
			sumOfIds = sumOfIds + gameId
		end
		local power = gameMax.red * gameMax.green * gameMax.blue
		print(
			"Game " .. gameId,
			"red:" .. gameMax.red,
			"green:" .. gameMax.green,
			"blue:" .. gameMax.blue,
			"power:" .. power
		)
		powerSum = powerSum + power
	end
	print("Sum of ids", sumOfIds)
	print("Sum of powers", powerSum)
end
Main()
